### ISSUE TITLE
"Problem Retrieving Data" and then crash

### ISSUE THREAD

**Issue by stefan2904** [signalapp/Signal-iOS#5811]
- [X] I have searched open and closed issues for duplicates
- [X] I am submitting a bug report for existing functionality that does not work as intended
- [X] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description

When I start Signal, I get the "Problem Retrieving Data" screen. After confirming it, the all displays that it is now recovering the database but crashes after 10 seconds. There is no way to skip the recovery or retrieve debug logs.

It might be the case that this was triggered by a full device storage, but now there is >1GB space free> 

### Steps to reproduce
- Start app
- Immediately get "Problem Retrieving Data" screen
- Click continue 
- For 10 seconds, get recovery screen

**Actual result:** Describe here what happens after you run the steps above (i.e. the buggy behaviour)

- App crashes without Error message or debug log

**Expected result:** Describe here what should happen after you run the steps above (i.e. what would be the correct behaviour)

- Successful Recovery 
- Able to start Signal and access messages 

### Screenshots

![IMG_20240520_105620_451](https://github.com/signalapp/Signal-iOS/assets/527767/c37a2404-a90f-4a0b-8d46-dd7486291fe7)

![IMG_20240520_105616_814](https://github.com/signalapp/Signal-iOS/assets/527767/c8b86ce4-b6d4-4d11-99a4-42ac5382ea31)

### Device info

**Device**: iPhone SE 2020

**iOS version**: 17.3.1

**Signal version:** 7.11

### Link to debug log

- Log from app not available 
- Log from iOS send to signal support.

**Comment by stefan2904** [comment]
The log indicates a `SIGTRAP` and some Address size faults below.

```
{"beforePC":"4gMZquMDF6rkAxWq5QMTquYDFqrnAxiqFQCA0pUDAJTgAxSq1xgPlA==","atPC":"IAAg1IgSgFL\/EwC56AcA+UgAgFLoAwA54CYAsADQLZHjJgCwY0A6kQ=="},
  "basebandVersion" : "5.00.00",
  "exception" : {"codes":"0x0000000000000001, 0x00000001a96fe3fc","rawCodes":[1,7137649660],"type":"EXC_BREAKPOINT","signal":"SIGTRAP"},
  "termination" : {"flags":0,"code":5,"namespace":"SIGNAL","indicator":"Trace\/BPT trap: 5","byProc":"exc handler","byPid":1265},
  "os_fault" : {"process":"Signal"},
  "faultingThread" : 2,
  "threads" :
```

**Comment by stefan2904** [comment]
fyi, the Signal support team replied to my email that they are aware of and working on that issue, but they cannot give a timeline.

So, in my case, I had to reset the app, losing all old messages. 😭

**Comment by acanis** [comment]
This exact same issue just happened to me. My iPhone was running out of space and I opened signal. Now after freeing up space the Signal app says "Problem Retrieving Data" and immediately crashes when I click continue. 
Is there no way to recover the data?

**Comment by birfan** [comment]
**Tap 8 times on the white area of "Problem Retrieving Data". It will ask "What would you like to do with the link to your debug log?", and choose email support.** I sent a bug report for the same issue last week. 

I just hope they will fix it soon. The more people send their report, the more they will (hopefully) pay attention to this bug. Last two updates haven't fixed it for me, and I have read every single post about this issue. There is no way to recover the data.

**Comment by Nerodelo** [comment]
I got the same issue, any update? I don't want to lose my data

**Comment by Nerodelo** [comment]
> **Tap 8 times on the white area of "Problem Retrieving Data". It will ask "What would you like to do with the link to your debug log?", and choose email support.** I sent a bug report for the same issue last week.
> 
> I just hope they will fix it soon. The more people send their report, the more they will (hopefully) pay attention to this bug. Last two updates haven't fixed it for me, and I have read every single post about this issue. There is no way to recover the data.

It's impossible for me to send report, app crash before

**Comment by birfan** [comment]
> > **Tap 8 times on the white area of "Problem Retrieving Data". It will ask "What would you like to do with the link to your debug log?", and choose email support.** I sent a bug report for the same issue last week.
> > I just hope they will fix it soon. The more people send their report, the more they will (hopefully) pay attention to this bug. Last two updates haven't fixed it for me, and I have read every single post about this issue. There is no way to recover the data.
> 
> It's impossible for me to send report, app crash before

Don't click "continue" on the "problem retrieving data" screen. Just tap 8 times on the white area. If that doesn't work, try contacting support directly.

**Comment by rocodes** [comment]
This may not be a helpful addition, but just a +1 that this is affecting another user I work with, who also had what seems like database corruption after a low storage warning.

I have advised them to share debug logs. Happy to assist otherwise if I can.

(Also https://github.com/signalapp/Signal-iOS/issues/5813, https://github.com/signalapp/Signal-iOS/issues/5815).

**Comment by mixmaxdp** [comment]
+1

I have the same issue - shared debug logs via email and really looking for the bug fix, since this is affecting core Signal functionality.

Please bump up the priority, happy to assist and share more logs/data from iOS if needed to streamline the fix.

**Comment by birfan** [comment]
This issue was fixed for me in the final update released yesterday (version 7.15)!

Make sure you have enough free storage on your phone (I had 6 GB), update the app, then open the Signal app. Continue with the database recovery step. It doesn't crash, the recovery is complete, and I got back both my old messages (including media shared) before the crash, and every message I received after. 

Thank you Signal team for fixing it!

**Comment by c-mar-coder** [comment]
I have encountered the same issue. My iPhone storage was full and I opened Signal and it crashed. Next time I opened Signal, it gave me the same error as described above ("Problem Retrieving Data"). When trying to retrieve data, it reaches 57% and then tells me that data cannot be retrieved and I am supposed to reset Signal completely which I do not want to do as I do not want to lose all my messages and media data. 
Interestingly enough I am still receiving new messages and they pop up on my home screen but I have no way to access them as I run into "problem retrieving data" again. 
Any new insights what I can do?

**Comment by benjamindahl** [comment]
exact same problem here. 
+1

**Comment by dronox** [comment]
Same issue here
+1

**Comment by hanifvirani** [comment]
I have the same exact issue. Got a storage warning on my iPhone and Signal gives me the "Problem Retrieving Data" screen ever since. I continue to receive notifications for new messages however. 

Looks like @birfan was able to do a successful recovery on v7.15 while @c-mar-coder's recovery failed. 

I haven't tried pressing "Continue" yet because I am afraid of losing all my data.

@c-mar-coder can you confirm you were able unable to fix this even after upgrading to v7.15?

Is there anyone else who got this error screen and was able to recover successfully?

**Comment by cosmicpudding** [comment]
This has just affected me as well - low storage, it glitched a bit and even though I cleared space and was receiving new messages, one “ghost” message never appeared and was causing a stuck notification. I tried updating the app to 7.32 to resolve it with no luck, and this morning I open it to see the exact messages as everyone here, and it fails to recover. I submitted my debug logs via the link, but right now am looking at seemingly losing my entire Signal history and all associated media which is pretty awful. 

If that recovery process is supposed to work then the only thing I can think of is to clear more space than I have (1.5 GB) in the hopes that maybe this is partly related to not enough space for the recovery process to work properly.

If that’s not the case then I’m pretty stuck until there is help from developers or a fix, which hopefully is possible…

**Comment by Adalfunz** [comment]
This happened to me on October 30th in the morning. I opened Signal and received the same message, "Problem Retrieving Data." I immediately continued with recovery but it got stuck at 56% for about 10 seconds, then continued to 72%, then the "Data Not Recovered" message came up. I figured all was lost so I went ahead and pressed "Rest Signal," and the app just crashed, so I went to Settings > General > iPhone Storage > Signal and it looked like all my data was still there. Documents & Data 11 GB. I was low on storage when this happened, so I deleted a bunch of apps. After sending the debug log, I received a response from the Signal team that I didn't have the latest update, so I updated the app. One thing I noticed was that my photos library kept restoring from iCloud so no matter how many apps I would delete or how many backups I would make, I kept losing space until the sync finished. On the 31st I got about 11 GB of storage space back, and that's when I updated iOS. So once I thought I was all set for recovery, I went ahead and pressed "Continue" again, but nothing changed. The only thing now is that I'm scared to press "Rest Signal" because perhaps this time I could actually lose my data due to more storage space. I don't know if that's possible. Maybe I'm wrong about that, but that's not a chance I'm willing to take. Hopefully there'll be a bug fix soon.

**Comment by hanifvirani** [comment]
In case this helps anyone else, I was able to successfully recover my data. I just made sure I had enough free storage space before clicking "Continue".

**Comment by OctopusET** [comment]
+1
I even made 25Gigabytes of free space, but didn't work. My signal data is around 4.5 Gigabyte

**Comment by mihaelacr** [comment]
Hitting the same issue: got messages while iphone storage space was full, can no longer restore data from signal (despite now having > 20GB free of space).

Looking at the logs:
```
2024/11/16 08:08:24:654  🧡 [DatabaseRecovery.swift:552 copyTable(tableName:from:to:)]: Error while iterating: SQLite error 11: database disk image is malformed - while executing `SELECT * FROM keyvalue`
2024/11/16 08:08:24:657  🧡 [DatabaseRecovery.swift:564 copyTable(tableName:from:to:)]: Error when reading: SQLite error 11: database disk image is malformed - while executing `COMMIT TRANSACTION`
2024/11/16 08:08:24:657  🧡 [DatabaseRecovery.swift:402 copyTableThatMustBeCopiedFlawlessly(tableName:oldDatabaseStorage:newDatabaseStorage:)]: Completely unable to copy keyvalue
```

Updating the app does not help the issue. 

As others have mentioned, the data is definitely still there (Signal is taking 11GB space on my phone), so there should be a way to fix the corrupt messages and restore the state for most data. 

Seems to be a duplicate of  [issue 5813](https://github.com/signalapp/Signal-iOS/issues/5813), which also has not resolution.

**Comment by trashcoder** [comment]
Same error here. My wife's iPhone had no empty space and signal crashed. After getting some space we got the same error as above.

**Comment by mihaelacr** [comment]
I have spent more time looking into this, and I believe the following steps would solve the problem:

1. allow the user to export the SQLlite database
2. the user fixes the database, for example via `sqlite3 mydata.db ".dump" | sqlite3 new.db`
3. allow the user to re-import the database.

1 and 3 can be done from the current failure UI screen. 2 is the responsibility of the user. 

I have reached out to the signal team but have not heard back on whether they are willing to implement steps 1 and 3.

In lieu of that, the next question to answer is whether if I build Signal from source myself (to implement 1), and put it on my iPhone, will it be able to see my data that was written by the Signal app from the app store?

**Comment by 1din555go** [comment]
> I have spent more time looking into this, and I believe the following steps would solve the problem:
> 
> 1. allow the user to export the SQLlite database
> 2. the user fixes the database, for example via `sqlite3 mydata.db ".dump" | sqlite3 new.db`
> 3. allow the user to re-import the database.
> 
> 1 and 3 can be done from the current failure UI screen. 2 is the responsibility of the user.
> 
> I have reached out to the signal team but have not heard back on whether they are willing to implement steps 1 and 3.
> 
> In lieu of that, the next question to answer is whether if I build Signal from source myself (to implement 1), and put it on my iPhone, will it be able to see my data that was written by the Signal app from the app store?

@mihaelacr I have the same problem, did you find any solution for this?

**Comment by nh2** [comment]
> if I build Signal from source myself (to implement 1), and put it on my iPhone, will it be able to see my data that was written by the Signal app from the app store?

I asked this [on iOSProgramming reddit](https://old.reddit.com/r/iOSProgramming/comments/1hd0ui3/possible_to_access_app_data_when_compiling_app/); they said one would **not** be able to get at one's own data this way.

Which is weird because it effectively means the data is effectively owned/controlled by the app developer (Signal devs), not by the "owner" of the Apple device.

Would be nice if anybody could confirm this.

**Comment by Nerodelo** [comment]
> > if I build Signal from source myself (to implement 1), and put it on my iPhone, will it be able to see my data that was written by the Signal app from the app store?
> 
> I asked this [on iOSProgramming reddit](https://old.reddit.com/r/iOSProgramming/comments/1hd0ui3/possible_to_access_app_data_when_compiling_app/); they said one would **not** be able to get at one's own data this way.
> 
> Which is weird because it effectively means the data is effectively owned/controlled by the app developer (Signal devs), not by the "owner" of the Apple device.
> 
> Would be nice if anybody could confirm this.

@nh2 Have you e-mailed your solution to the signal team? No update has solved the problem since last year

**Comment by nh2** [comment]
@Nerodelo Which solution? According to my [post above](https://github.com/signalapp/Signal-iOS/issues/5811#issuecomment-2557874726), there's no solution, because the Signal team owns our data with their signing key, and we as users (and developers) can't get to it.

The only real solution seems to be if the Signal team builds a raw DB export feature into the main Signed signal app, or provides us with a signed app out-of-band that has this feature.

**Comment by Blaualgen** [comment]
I still have the same problem, I'm super sad I could loose years of conversation 
Did anyone succeed with an iMazing (Im using iPhone7, iOS 15.7.1) encrypted backup to restore any part of it?

**Comment by kurtislamtk** [comment]
> I still have the same problem, I'm super sad I could loose years of conversation Did anyone succeed with an iMazing (Im using iPhone7, iOS 15.7.1) encrypted backup to restore any part of it?

Same problem. That's super sad. I'm also thinking of some external help but I didn't see any example which is successful from others.

**Comment by 1din555go** [comment]
I feel the same. I also have the same problem. It’s so sad and it’s not even something I caused. It wasn't my fault. Still no reply from support.

**Comment by kevin-valerio** [comment]
Same problem.

**Comment by sashaweiss-signal** [comment]
Hi, we're very sorry you're experiencing these issues. Database corruption is a longstanding and very complex issue that we've made progress on, but is obviously still present.

For anyone crashing during the "Database Recovery" flow, it'd be super valuable if you could provide `.ips` crash files so we can investigate. Instructions for that can be found [here](https://support.signal.org/hc/en-us/articles/360007318591-Debug-Logs-and-Crash-Reports#ios_debug).

Additionally, if you're able to collect debug logs (either from the sheet that pops during launch saying something is wrong, or by tapping 8 times anywhere on the screen), those would be helpful in addition to the `.ips` files.

**Comment by kurtislamtk** [comment]
Hi @sashaweiss-signal , thanks.

Please find the debugging log here: https://debuglogs.org/ios/7.76.0/347378d88424cabc271262f03bf086a1fb1656554ec6bc821690274d39e23f6d.zip

I cannot upload the .ips file here because Github said it is not supported to upload. Would you advise if there is any other means that I could send the file to you? Thanks.

**Comment by sashaweiss-signal** [comment]
Thanks for that. To upload the `.ips` file, can you change the file extension from `.ips` to `.zip` or even `.png` (without editing the file)? Github will allow the upload then, and I can adjust it back to `.ips` once I've gotten it.

**Comment by kurtislamtk** [comment]
[Signal-2025-09-11-234712.ips.zip](https://github.com/user-attachments/files/22280364/Signal-2025-09-11-234712.ips.zip)

Done. Many thanks for your help! @sashaweiss-signal

**Comment by sashaweiss-signal** [comment]
Thanks for that @kurtislamtk. I'm sorry to report that the logs and `.ips` file suggest that the database is unrecoverably corrupted.

That said, we've made a couple improvements to the database recovery flow, which are available now in the 7.77 beta release (see [here](https://community.signalusers.org/t/beta-feedback-for-the-upcoming-ios-7-77-release/71585/9) for more details). As a last-ditch effort, I'd suggest upgrading to 7.77 (either while it's in beta, or once it goes to production next week) and seeing if you're able to go through database recovery.

Again, I'm very sorry this is happening, and thank you for helping us understand the problem better.

**Comment by fschwar4** [comment]
@sashaweiss-signal Could you have a look at these files? (they are with the new version 7.77; push notification sometimes still work which is strange to me).

Thank you in advance!!

[0997f0c92643071fc4d90a290deb6dc3ca0865169238dd8040c64255a2d93ed1.zip](https://github.com/user-attachments/files/22502793/0997f0c92643071fc4d90a290deb6dc3ca0865169238dd8040c64255a2d93ed1.zip)

**Comment by Virvlgames** [comment]
I've encountered the same issue, here is my debug log:

[cec5ee98a47ed393077fe88a6e5539bb1bada50612fc656d15e36e41c2e00edf.zip](https://github.com/user-attachments/files/22525547/cec5ee98a47ed393077fe88a6e5539bb1bada50612fc656d15e36e41c2e00edf.zip)

**Comment by sashaweiss-signal** [comment]
Hi @fschwar4, @Virvlgames, thanks for submitting those logs.

I'm very sorry to report that in both cases the Database Recovery tool was unable to complete, because of corruption in a database table we consider critical. Consequently, despite our best efforts your databases appear unrecoverable, and your only option to continue using Signal will be to uninstall and reinstall.

**Comment by trashcoder** [comment]
What about checking the free disk space before signal makes a write transaction to its database?

**Comment by FDMS** [comment]
Hi @sashaweiss-signal, I would _really_ appreciate it if you could check mine as well:
- https://debuglogs.org/ios/7.79.0/6d64bff22ce14f2bdbdcd429ddff8fd24ab3414df609c8fece8c6c8e27b06657.zip
- [Signal.cpu_resource-2025-09-28-090836.ips.zip](https://github.com/user-attachments/files/22580846/Signal.cpu_resource-2025-09-28-090836.ips.zip)

(@ Everone else: Sorry for the notification)

**Comment by fschwar4** [comment]
@sashaweiss-signal thank you for checking! Would it be any good if I install an Iphone Backup from before the loss?
Thank you in advance!

**Comment by sashaweiss-signal** [comment]
> What about checking the free disk space before signal makes a write transaction to its database?

We have some ideas along these lines, but they all come with tradeoffs that we're exploring.

> check mine as well

@FDMS, unfortunately similarly your logs show that the `keyvalue` table is corrupted, which is a critical table. We're looking into options here, such as storing less data in `keyvalue` to reduce the likelihood of it being corrupted, but I'm very sorry to report that for now it looks like it is critically corrupted.

```log
2025/09/28 07:23:18:175  WRN🧡 [DatabaseRecovery:566 copyTable]: Error while iterating: SQLite error 11: database disk image is malformed - while executing `SELECT * FROM keyvalue`
2025/09/28 07:23:18:181  WRN🧡 [DatabaseRecovery:578 copyTable]: Error when reading: SQLite error 11: database disk image is malformed - while executing `COMMIT TRANSACTION`
2025/09/28 07:23:18:181  WRN🧡 [DatabaseRecovery:414 copyTableThatMustBeCopiedFlawlessly]: Completely unable to copy keyvalue
```

> Would it be any good if I install an Iphone Backup from before the loss?

Signal is not included in standard iCloud iPhone Backups, so unfortunately not.

**Comment by ecalifornica** [comment]
> > What about checking the free disk space before signal makes a write transaction to its database?
> 
> We have some ideas along these lines, but they all come with tradeoffs that we're exploring.
> 

I'd be curious to hear about these tradeoffs.


> Signal is not included in standard iCloud iPhone Backups, so unfortunately not.

The Signal db doesn't seem to be included in the local backups either.

**Comment by XP-Luna** [comment]
I have the same issue right now since yesterday… Someone sent me a photo, after that this happened. Before it was working. I got that message that I have to recover before so I thought it’s no big deal, it was able to recover no problem the last time. However this time it wasn’t… I can’t lose everything in my Signal history, it’s such a shame the backup feature isn’t out yet, I wanted to turn that on day one… Everything is there as it is my main communication app, so many photos and so on, I need it back. Also I already got a new phone that runs iOS 26 since a few days, I wanted to transfer it there this week… What a bad timing.

Honestly I would even pay you to recover this manually, this is maybe 10 years of memories and important conversations and I just had no way to back it up, this is so bad… I mean the data is still there obviously and I really can’t afford to lose all of it.

Thanks in advance

https://debuglogs.org/ios/7.82.0/61a44fde012664a8b1107363b74d5048d66ac415dee97b11956a342c1781ae24.zip

**Comment by XP-Luna** [comment]
Also I could jailbreak that phone to get access to the raw files as it is running iOS 15.0 if that helps and if a manual recovery could be possible then for example

**Comment by andrade** [comment]
This happened to me today on an iPhone SE 1st gen.

- The first log I sent I got a reply with _This is not the latest version of Signal._.
- The second log I sent, after updating the app (from .84 to .85), I got a reply with _It appears the database is irrecoverably damaged. Unfortunately, you will need to completely uninstall and then reinstall Signal. This will delete your message history. It cannot be recovered._.

In both cases the reply was very quick (automated I would guess). I end up hitting reset instead of reinstalling and it appears to be working now but all my chats are gone from the phone. :-\  (I had it linked with desktop/laptop so parts are still there including media.)

**Comment by XP-Luna** [comment]
I got the same mail within a minute, so I’m pretty sure it is automated which is honestly fine for the first email in which you explain a known issue.

I have not reset it as I really can’t afford to lose the massages and media, there is some very sentimentally important but also otherwise important messages in there that I need to have back. As I said, the data is clearly still there, one database got corrupted that it can’t rebuild on its own (yet). I hope there will be an update that hopefully introduces the ability to recover this type of database error and will fix it, I’m willing to wait a bit for that. Or, as I said, I would be down for manual repairing together with a dev or sending the phone to them. I would even pay / donate for that

**Comment by Jannel-r** [comment]
This happened to me today. My internal storage space was almost full and when a friend sent me a video, signal would not open. I have since cleared internal storage space >3GB free now and it still won't work. After I try opening the app several times I get a message that reads:
"Couldn't Open Signal
Signal had a problem while trying to open. Please send a debug log to support@signal.org to help troubleshoot this issue."
The first time I submitted the Debug log I got an automated reply that said "this is not the latest version of Signal"
I went to the app store and updated to the latest version of Signal and the problem persisted. I sent another Debug report and was told "it appears the database was irrecoverably damaged" and to uninstall and reinstall which will result in me losing all my conversations. I am still receiving messages (the notifications pop up) but have no way to access them or open Signal.

@sashaweiss-signal can you pretty please look at the debug log and .ips below and tell me if there is any hope? I would really super appreciate it!

--- Support Info ---
Filter: Signal - iOS Debug Log - LaunchFailure_LastAppLaunchCrashed
Challenge Received: no
Hardware: iPhone
HID: 9,3
iOS Version: 15.8.5 (19H394)
Signal Version: 7.86.2.1206
Log URL: https://debuglogs.org/ios/7.86.2/65cfe63bdbe42ec118d96157927f3097804f7301e5bade38bc12f1ee7a14e8a6.zip
Locale: en_CA

[Signal-2025-11-28-090116.ips.zip](https://github.com/user-attachments/files/23829565/Signal-2025-11-28-090116.ips.zip)

[ReportCrash-2025-11-28-090155.ips.zip](https://github.com/user-attachments/files/23829568/ReportCrash-2025-11-28-090155.ips.zip)

I have a bunch of .ips files but the 090116 was the latest. There are also SignalNSE .ips reports. I can send all the reports if they would be helpful for Signal to figure out a solution to this problem.

**Comment by C-Duv** [comment]
The screen that says Signal can't load the database and offers to send debug logs should offer to retrieve/download/share the database (as a plain SQLite file).

At least we could then try to recover it with SQLite tools on our computer.

**Comment by XP-Luna** [comment]
Unfortunately it doesn’t give me that option as you can see, but if extracting the raw SQLite database was possible that could be huge for trying to recover this. I also updated to the latest version of Signal but the option still isn’t there. Or am I not seeing something very obvious?

Thank you so much though for giving us some hope and taking the time for this issue.

<img width="1170" height="2532" alt="Image" src="https://github.com/user-attachments/assets/4044d52d-561d-4862-a1aa-eddc1a27799b" />

**Comment by C-Duv** [comment]
My fault @XP-Luna: I did not meant the screen "should" offer to get the database as in "it's actually coded to do so and you should be able to see the feature".

As I also have the issue on my phone, I was asking (the devs) if the "failed to …" screen could be modified to have an option to download/get the database, because I think it's the more portable way to unblock users like us.

**Comment by XP-Luna** [comment]
Oh okay, yeah that seems to be a good idea honestly.

**Comment by C-Duv** [comment]
@XP-Luna I've looked at the code and a DB extraction seems feasible, I've posted my findings in #5815: https://github.com/signalapp/Signal-iOS/issues/5815#issuecomment-3627185722 but did not got any traction (yet?).

**Comment by mlthrk** [comment]
Unfortunately, my database was also corrupted, resulting in the possible loss of many messages with my deceased father. Is there any way to save/export the corrupted database from my phone to hopefully repair it in the future? Is the database included in local encrypted backups?

**Comment by XP-Luna** [comment]
@C-Duv That really seems like a good solution, I mean the whole pitch of Signal is that we own or own personal data, so why not allow exporting the locally stored database stored on our own phone with our data

**Comment by mlthrk** [comment]
Oddly enough, I started to receive notifications from new messages a few days ago. They seem to be messages from after the database was corrupted. Still no luck opening Signal, though. Is there still hope to recover it?

**Comment by C-Duv** [comment]
> [@C-Duv](https://github.com/C-Duv) That really seems like a good solution, I mean the whole pitch of Signal is that we own or own personal data, so why not allow exporting the locally stored database stored on our own phone with our data

According to "5_c_d" forum user on [this "Unable to View database of iOS app in DBHelper or SQLiteManager" thread](https://community.signalusers.org/t/unable-to-view-database-of-ios-app-in-dbhelper-or-sqlitemanager/7865/2), database is encrypted:

> The database is encrypted, and the iPhone secure-element is used to store the encryption-keys.

I don't know how's the encryption done, but it might (or not) means extraction of the raw .sqlite file is pointless, you would get a encrypted file whose key is out of reach…

**Comment by mihaelacr** [comment]
This (the issue from the original post) has happened to me again... I cannot believe this is still an issue and that the option to recover the SQLite database is not given to the user.

**Comment by XP-Luna** [comment]
A few updates ago the screen when opening Signal has changed. Now there is the option to open Signal instead? What does that button do if the database is not repaired but it’s also not reset? 

I did not tap it yet because I don’t want to kill my data. Also the progress when trying to repair now goes way further than before. Instead of 53 % ish ist goes to like 76 % before failing. Has something changed there?

Now secure backups are finally here, great. This means in the future it’s actually possible for people to take matters into their own hands and avoid losing all of their message history, pictures and videos. If they decide to turn it on. Thank you for implementing that, I wish I could have used it a few months ago. However for the people who are in this situation right now a solution remains to be seen. As we know, the data itself is still there, just the database is corrupted which should be fixable, even if only manually. This is my current log: https://debuglogs.org/ios/8.0.0/0ec36f040653999c23fd77f34498a7fc6f3fa53ef4590b6b0e9b346f842e2b53.zip

<img width="1170" height="2532" alt="Image" src="https://github.com/user-attachments/assets/b23dba5a-9887-4be8-aa7f-bb1d89793e79" />

**Comment by max-signal** [comment]
> Now there is the option to open Signal instead? What does that button do if the database is not repaired but it’s also not reset?

If we get back a `SQLITE_CORRUPT` error from SQLite, we mark the database as corrupt and show this interface. That corruption might be at the very “beginning” of the database, or it may be buried deep in some old conversation or message. It’s not always easy to know whether or not trying again will need to access the corrupt data or not.

A while back (in a237b9c1148860978d0314809702c0dbd7768628), we made some changes here to always allow you to bypass the repair step and try to use the database in its current form. It’s likely that you’ll run into corruption again and get stuck, but it’s possible you won’t. Tapping “Open Signal Instead” won’t intentionally delete anything.

> As we know, the data itself is still there, just the database is corrupted which should be fixable, even if only manually.

When the corruption triggers our checks, it normally means that something about the structure of the database is malformed, and we may not be able to distinguish “this is part of message 1” from “this is message 2” (or even “where are the messages?”). You may be able to guess at how the database is supposed to look, but there’s not really a generalized way to recover these databases. (We try to recover what we can and throw away state that can be recreated, and that’s why repair sometimes works and sometimes doesn’t.)

**Comment by XP-Luna** [comment]
Thanks for the detailed explanation, I really appreciate it. However for me pressing Open Signal Instead just instantly closes the app, presenting me with the same screen when reopening…

**Comment by C-Duv** [comment]
> When the corruption triggers our checks, it normally means that something about the structure of the database is malformed, and we may not be able to distinguish “this is part of message 1” from “this is message 2” (or even “where are the messages?”). You may be able to guess at how the database is supposed to look, but there’s not really a generalized way to recover these databases. (We try to recover what we can and throw away state that can be recreated, and that’s why repair sometimes works and sometimes doesn’t.)

Could you give us the opportunity to try to repair or salvage it on a computer?

**Comment by redm123** [comment]
This is so ridiculous! Something as basic as running out of storage leads to data corruption. And the only "solution" is to throw away all my chat history! This must be a joke. But it isn't. For years I cannot even make a backup, into my standard encrypted iPhone backup, to some other local system, whatever. I cannot restore the history from my Signal Desktop. Nothing. Because it's oh so insecure ... (and the first thing they offer is a cloud backup ... for 45 days... more ridiculous)
This app is really just useless crap for throw away messaging. If you want something serious, use something else. I can't stop shaking my head...

**Comment by arnoutj** [comment]
Same problem here, would appreciate if we could still extract the corrupt db

**Comment by kritschy** [comment]
Same problem here. Signal crashed and DB couldn’t be retrieved. @devs I think it would be insanely important to present a solution for this;  i fear its really bad PR bc lots of people are affected. Especially in combination with those phishing attacks  :///

**Comment by MrParanoid42** [comment]
A friend of mine is having the same issue as above. The only person they speak to is myself for the most part, is there a way to use my database to rebuild theirs?

**Comment by johnw188** [comment]
I dug into the debug logs from a fresh instance of this (Signal 8.10.0 build 1570, iOS 18.1.1, iPhone13,3) and the root cause is not disk space — it's how `DatabaseRecovery` handles the `keyvalue` table. This is the exact failure @mihaelacr posted above, and likely explains most of the "recovery reaches ~50–75% then fails" reports in this thread and in #5813.

**What actually happens**

Repair's best-effort pass copies every table successfully — 129,701 interactions, 4,707 threads, ~12k attachments, all recipients/profiles/groups. Then it hits `keyvalue`, which is copied via the "flawless" path, and a single corrupted page kills the cursor mid-iteration. The entire recovered DB — including all the messages that copied cleanly — is deleted. Relevant excerpt from my log:

```
[DatabaseRecovery:748 integrityCheck]: [DatabaseRecovery] Integrity check failed
[DatabaseRecovery:43 reindex]: [DatabaseRecovery] Attempting to reindex the database...
[DatabaseRecovery:58 reindex]: [DatabaseRecovery] Failed to reindex database
...
[DatabaseRecovery:389 copyWithBestEffort]: [DatabaseRecovery] Finished copying model_TSInteraction. Copied 129701 row(s)
[DatabaseRecovery:389 copyWithBestEffort]: [DatabaseRecovery] Finished copying model_TSThread. Copied 4707 row(s)
[DatabaseRecovery:389 copyWithBestEffort]: [DatabaseRecovery] Finished copying Attachment. Copied 12257 row(s)
[DatabaseRecovery:389 copyWithBestEffort]: [DatabaseRecovery] Finished copying MessageAttachmentReference. Copied 13236 row(s)
...
[DatabaseRecovery:443 copyTableThatMustBeCopiedFlawlessly]: [DatabaseRecovery] Attempting to copy keyvalue (with no mistakes)...
[DatabaseRecovery:626 copyTable]: [DatabaseRecovery] Error while iterating: SQLite error 11: database disk image is malformed - while executing `SELECT * FROM keyvalue`
[DatabaseRecovery:638 copyTable]: [DatabaseRecovery] Error when reading: SQLite error 11: database disk image is malformed - while executing `COMMIT TRANSACTION`
[DatabaseRecovery:451 copyTableThatMustBeCopiedFlawlessly]: [DatabaseRecovery] Completely unable to copy keyvalue
[DatabaseRecovery:428 prepareToCopyTablesThatMustBeCopiedFlawlessly]: Couldn't copy tables flawlessly: SQLite error 11: Optional("database disk image is malformed") (extended result code: 11)
[DatabaseRecovery:225 deleteTemporaryDatabase]: [DatabaseRecovery] Attempting to delete temporary database files...
[DatabaseRecovery:234 deleteTemporaryDatabase]: [DatabaseRecovery] Deleted temporary database file
```

That's why freeing storage doesn't help and why Repair fails identically on every retry — it's deterministic on the same corrupted page in one non-message table.

(In my case the corruption is old: `lastCompletedLaunchMainAppVersion: 7_74_0_972` vs `lastCompletedLaunchNSEAppVersion: 8_10_0_1570` — the main app hadn't completed a launch across many versions while the NSE kept running, so this likely happened during a background migration and sat undetected.)

**Why it's fixable**

Both copy paths use the same row-by-row `copyTable()`, and rows copied before the cursor dies are already committed to the new DB. So in the `.copiedSomeButHadTrouble` case the partial `keyvalue` data is *already sitting in the recovered database* — the code just throws it away:

```swift
case let .totalFailure(error), let .copiedSomeButHadTrouble(error, _):
    Logger.warn("Couldn't copy tables flawlessly: \(error.grdbErrorForLogging)")
    let toThrow: DatabaseRecoveryError = error.isSqliteFullError ? .ranOutOfDiskSpace : .unrecoverablyCorrupted
    throw toThrow
```

Losing some KV rows (worst case: re-registering the number) is a strictly better outcome for the user than "Reset App Data" discarding a message history that had, in fact, just been fully recovered. Same class of problem as #6287.

**Proposed patch** (`SignalServiceKit/Storage/Database/DatabaseRecovery.swift`, +16/−1) — accept `.copiedSomeButHadTrouble` for `keyvalue` when at least one row copied; keep `.totalFailure` and disk-full as hard failures:

```diff
                     switch result {
-                    case let .totalFailure(error), let .copiedSomeButHadTrouble(error, _):
+                    case let .totalFailure(error):
                         Logger.warn("Couldn't copy tables flawlessly: \(error.grdbErrorForLogging)")
                         let toThrow: DatabaseRecoveryError = error.isSqliteFullError ? .ranOutOfDiskSpace : .unrecoverablyCorrupted
                         throw toThrow
+                    case let .copiedSomeButHadTrouble(error, rowsCopied):
+                        Logger.warn("Couldn't copy tables flawlessly: \(error.grdbErrorForLogging)")
+                        if error.isSqliteFullError {
+                            throw DatabaseRecoveryError.ranOutOfDiskSpace
+                        }
+                        // A single corrupted page in keyvalue can fail the cursor mid-iteration
+                        // even when most rows (and every other table) copied cleanly. Rather
+                        // than abort an otherwise-successful recovery, keep the partial copy:
+                        // rows that did copy are already committed by `copyTable`.
+                        if tableName == KeyValueStore.tableName, rowsCopied > 0 {
+                            Logger.warn("Falling back to best-effort copy of \(tableName); copied \(rowsCopied) row(s). Some keyvalue rows may have been lost.")
+                            continue
+                        }
+                        throw DatabaseRecoveryError.unrecoverablyCorrupted
                     case .wentFlawlessly:
                         break
                     }
```

`DatabaseRecoveryTest.testDumpAndRestoreWithInvalidEssentialTable` (drops `keyvalue` entirely → `.totalFailure` → still throws) passes unchanged. Happy to open a PR with this, or with a more targeted variant (e.g. explicitly re-seeding the launch-gating KV entries after a partial copy) if maintainers would prefer to keep the flawless invariant strict.

**Comment by XP-Luna** [comment]
This sounds really promising, thank you very much for that detailed investigation. I really hope a fix based on that can be implemented soon so we can recover our years of message history, images and videos again. I really appreciate your effort!

**Comment by nh2** [comment]
> the root cause is not disk space

> and a single corrupted page kills the cursor

@johnw188 What causes the corrupted page if not lack of disk space when writing?

**Comment by redm123** [comment]
Sounds to me as if there are two issues here: 1st the real problem: that low disc space leads to corruption in the first place. This should be handled before it fails and never happen! 2nd what the investigation above describes: if it happens, at least gracefully recover...  
But really number 1 needs to be fixed!

**Comment by johnw188** [comment]
@nh2 you're totally right, my apologies. I'm helping out a friend who's locked out of their message history, I've been looking at the problem from the unable to recover angle not the cause of the corruption.

**Comment by bederke** [comment]
@max-signal @sashaweiss-signal

https://github.com/signalapp/Signal-iOS/issues/5811#issuecomment-5018049527

Please have a look at this - it's a big problem for me at the moment, as I really need Signal right now to communicate with a few groups and I don't want to lose my messages by resetting the app. 🥲 It’s a shame that something like this happens at all with a messaging app. 🫠

**Comment by XP-Luna** [comment]
Same for me, I really need it back and don’t want to lose it…

**Comment by flohansen1** [comment]
@max-signal  @sashaweiss-signal 

Hi i have a question, i am confused about what i should do. I had the same problem as descriebed above my i phone didn't have enough storage and then signal said database is corrupted. I tried to repair it unfortunatly before i freed up enough space and then i did it again later. Now i updated signal to the latest version and i now have 14GB of free storage my signal uses 10.2 GB of storage. Should i try to recover again or could it make the problem worse? I really need my old messages back. Or should i try to open signal istead or could this delete anything. Should i try to repair again and if it fails send you the report? Thank you for your help!

**Comment by Gitobont** [comment]
(I'm copying my comment from another ticket here, as they are very similar and I'm not sure which one is most relevant.)


Same thing happened to me. iOS device almost full, local Signal accepting messages > corrupt unrecoverable database.

Utterly unbelievable that the Signal app is allowed to DESTRUCTIVELY fail in this manner. What happened to graceful degradation of service ? And how is this still a bug after all these years ?!

I'm torn between the good feelings I have for Signal (years of private conversations, for free!), and this monumental screw-up. I'm not sure I will keep using it.

And it wouldn't be just me leaving : I have convinced all my friends and family to use Signal (I never used WhatsApp), and now that I'm not using it any more they very well might leave too. (Lots of them have told me that I am their only contact using Signal).

----------

Is there any chance of seeing a fix that will allow us to at least recover some messages and media ?

Can we get a comment from the Signal team on @johnw188 's [idea above](https://github.com/signalapp/Signal-iOS/issues/5811#issuecomment-5018049527) ?

**Comment by Gitobont** [comment]
In an effort to stay positive and constructive, some questions to whomever has more info : 

Is there any chance of a solution for recovering some of the content from the corrupted database ?
Would it be wise to switch phones and to stop using the problematic iPhone until said solution is offered ?
Would it be advisable to stop using Signal altogether until that solution is offered (or at least a definitive answer) ?

**Comment by omdathetkan** [comment]
Also running into this... here's a summary generated by Claude after analyzing my crash logs:

> Same issue here (#5811), and my logs match the .copiedSomeButHadTrouble case @johnw188 described.
> 
> Device: iPhone 12 mini (iPhone13,1), iOS 18.6.1 (22G90)
> Signal: 8.27.0.1843
> Trigger: device ran out of storage (~200 MB free) while the NSE was writing an incoming message → SQLITE_CORRUPT. Space has since been freed (multiple GB), which did not help.
> 
> Database Recovery copies everything cleanly, then fails only on keyvalue:
> 
> So this is not the error 11 / total read failure variant — the keyvalue rows read fine (5121 copied), it's a duplicate-key UNIQUE constraint violation, i.e. exactly the partial-copy case @johnw188's patch is meant to keep instead of discarding an otherwise-fully-recovered DB. All my messages copied cleanly and are being thrown away.
> 
> "Open Signal Instead" lets the app launch, but enabling Secure Backups crashes at the "view recovery key" step (it needs to write into the corrupt keyvalue table), so I have no way to back up before resetting.
> 
> Happy to send debug logs and .ips crash files. Please consider merging the partial-keyvalue recovery fix — for this UNIQUE-constraint variant the full message history appears intact and recoverable.

**Comment by omdathetkan** [comment]
@sashaweiss-signal  - could you please advise if there is any change that this will be fixed. Trying to decide on if I should bite the bullet and delete 5y worth of messages or hold on to a non-working Signal installation a little longer
